lib/psych/visitors/yaml_tree.rb

```diff
@@ -310,19 +310,20 @@ def visit_String o
           style = Nodes::Scalar::LITERAL
           plain = false
           quote = false
-        elsif o =~ /\n/
+        elsif o =~ /\n[^\Z]/  # match \n except blank line at the end of string
           style = Nodes::Scalar::LITERAL
         elsif o == '<<'
           style = Nodes::Scalar::SINGLE_QUOTED
           tag   = 'tag:yaml.org,2002:str'
           plain = false
           quote = false
+        elsif o.length > line_width
+          style = Nodes::Scalar::FOLDED
+          o << "\n" unless o =~ /\n\Z/  # to avoid non-default chomping indicator
```

Comment:
Why is the newline being added anyway?  I don't really understand it.  It doesn't seem good because it means that strings won't correctly roundtrip:

```ruby
    def test_line_width_roundtrip
      xs = 'xxx xxx' * 30
      xz = xs.dup # add the dup to avoid side effects
      yaml = Psych.dump xs, line_width: 12
      assert_equal xz, Psych.load(yaml)
    end
```